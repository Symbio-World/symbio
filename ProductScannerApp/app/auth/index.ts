import auth from '@react-native-firebase/auth'
import { createAuthProvider } from './auth-context'

export { useAuth, User } from './auth-context'

export const AuthProvider = createAuthProvider({
  signInAnonymously: async () => {
    const { user } = await auth().signInAnonymously()
    return { id: user.uid }
  },
})
