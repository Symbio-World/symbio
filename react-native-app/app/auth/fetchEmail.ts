import { firestore } from '../firestore'
import { EventType } from '@symbio/event-store-core'

export type FetchEmail = (userId: string) => Promise<string | undefined>

export const fetchEmail: FetchEmail = async (userId) => {
  const snapshot = await firestore()
    .collection(EventType.USER_SUBMITTED_EMAIL)
    .where('userId', '==', userId)
    .orderBy('timestamp', 'desc')
    .limit(1)
    .get()

  return snapshot.docs.length > 0 ? snapshot.docs[0].data().email : undefined
}
