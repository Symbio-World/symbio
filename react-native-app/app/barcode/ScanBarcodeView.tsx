import * as React from 'react'
import {
  AppState,
  AppStateStatus,
  View,
  Platform,
  PermissionsAndroid,
  BackHandler,
  Text,
} from 'react-native'
import { BarcodePicker } from 'scandit-react-native'
import { t } from 'react-native-tailwindcss'
import { settings } from './ScanBarcodeViewSettings'
import { Scanner, Session } from './Scanner'
import { ScanBarcodeScreenNavigationType } from './ScanBarcodeScreen'

type Props = {
  onScan?: (barcode: string) => void
  isActive?: boolean

  // a hack to test scanner ref as createNodeMock is not working https://github.com/callstack/react-native-testing-library/issues/227
  useRef?: typeof React.useRef
  navigation: ScanBarcodeScreenNavigationType
}

const isAndroidMarshmallowOrNewer = () => {
  return Platform.OS === 'android' && Platform.Version >= 23
}

const hasCameraPermission = () =>
  isAndroidMarshmallowOrNewer()
    ? PermissionsAndroid.check(PermissionsAndroid.PERMISSIONS.CAMERA)
        .then((isGranted) => isGranted)
        .catch(() => false)
    : Promise.resolve(true)

const requestCameraPermission = () => {
  return new Promise((resolve, reject) => {
    if (isAndroidMarshmallowOrNewer()) {
      PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.CAMERA)
        .then((granted) => {
          if (granted === PermissionsAndroid.RESULTS.GRANTED) {
            console.log('Android Camera Permission has been granted.')
            return resolve()
          }
          console.log(
            'Android Camera Permission has been denied - the app will shut itself down.',
          )
          return reject()
        })
        .catch((err) => {
          console.warn(err)
          reject()
        })
      return
    }
    resolve()
  })
}

const checkForCameraPermission = () => {
  return new Promise((resolve, reject) => {
    hasCameraPermission()
      .then((hasPermission) => {
        if (hasPermission) {
          resolve()
          return
        }
        requestCameraPermission().then(resolve).catch(reject)
      })
      .catch(reject)
  })
}

export const ScanBarcodeView: React.FC<Props> = ({
  onScan = () => {},
  isActive = true,
  useRef = React.useRef,
  navigation,
}) => {
  const scanner = useRef<Scanner>(null)
  const [isScannerInitialize, setScannerInit] = React.useState(false)

  const startScanning = () => {
    try {
      scanner.current?.startScanning()
      setScannerInit(true)
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  const resumeScanning = () => {
    try {
      scanner.current?.resumeScanning()
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }
  const pauseScanning = () => {
    try {
      scanner.current?.pauseScanning()
    } catch (e) {
      console.log('try-catch introduced so that simple render tests pass')
    }
  }

  React.useEffect(() => {
    const handleAppStateChange = (nextAppState: AppStateStatus) => {
      if (nextAppState.match(/inactive|background/)) {
        pauseScanning()
      } else {
        checkForCameraPermission()
          .then(startScanning)
          .catch(BackHandler.exitApp)
      }
    }

    AppState.addEventListener('change', handleAppStateChange)
    checkForCameraPermission().then(startScanning).catch(BackHandler.exitApp)
    return () => {
      AppState.removeEventListener('change', handleAppStateChange)
    }
  }, [navigation])

  React.useEffect(() => {
    if (!isScannerInitialize) return
    if (isActive) {
      resumeScanning()
      return
    }
    pauseScanning()
  }, [isActive, isScannerInitialize])

  const handleScan = (session: Session) => {
    const barcode = session.newlyRecognizedCodes[0].data
    onScan(barcode)
  }

  return (
    <View style={[t.flex1]}>
      <BarcodePicker
        onScan={handleScan}
        scanSettings={settings}
        ref={scanner}
        style={[t.flex1]}
      />

      <View
        style={[
          t.absolute,
          t.bottom0,
          t.wFull,
          t.h16,
          t.justifyCenter,
          t.bgBlack,
          t.opacity50,
        ]}
      >
        <Text style={[t.textWhite, t.textCenter, t.fontBold, t.text3xl]}>
          Scan barcode
        </Text>
      </View>
    </View>
  )
}
