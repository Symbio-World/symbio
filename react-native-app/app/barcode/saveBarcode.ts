import { userScannedBarcode } from '@symbio/event-store-core'
import { storeEvent } from '../event'

type SaveBarcode = (userId: string, barcode: string) => Promise<void>
export const saveBarcode: SaveBarcode = (userId, barcode) =>
  storeEvent(userScannedBarcode(userId, barcode))
