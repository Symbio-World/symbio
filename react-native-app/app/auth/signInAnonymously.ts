import auth from '@react-native-firebase/auth'

export const signInAnonymously = async () => {
  const { user } = await auth().signInAnonymously()
  return { id: user.uid }
}
