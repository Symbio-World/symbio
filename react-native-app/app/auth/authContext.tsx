import * as React from 'react'
import { FetchEmail } from './fetchEmail'

export type User = {
  id: string
  email?: string
}

type AuthContext = {
  user?: User
  setEmail: (email: string) => void
}

const authContext = React.createContext<AuthContext>({
  setEmail: () => {},
})

type Props = {
  children: React.ReactNode
}
export type SignInAnonymously = () => Promise<User>
export type Deps = {
  signInAnonymously: SignInAnonymously
  fetchEmail: FetchEmail
}
type CreateAuthProvider = (deps: Deps) => React.FC<Props>
export const createAuthProvider: CreateAuthProvider = ({
  signInAnonymously,
  fetchEmail,
}) => ({ children }: Props) => {
  const [user, setUser] = React.useState<User>()

  React.useEffect(() => {
    signInAnonymously().then((user) => {
      fetchEmail(user.id).then((email) => {
        if (email) {
          setUser({ ...user, email })
        } else {
          setUser(user)
        }
      })
    })
  }, [signInAnonymously, fetchEmail])

  const setEmail = (email: string) => {
    if (user) {
      setUser({ ...user, email })
    }
  }

  return (
    <authContext.Provider value={{ user, setEmail }}>
      {children}
    </authContext.Provider>
  )
}

export const useAuth = (): AuthContext => React.useContext(authContext)
