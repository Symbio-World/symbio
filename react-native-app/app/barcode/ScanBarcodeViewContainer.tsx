import * as React from 'react'
import { useAuth } from '../auth'
import { ScanBarcodeView } from './ScanBarcodeView'
import { saveBarcode } from './saveBarcode'
import { ScanBarcodeScreenNavigationType } from './ScanBarcodeScreen'

type Props = {
  isActive?: boolean
  onScan?: (barcode: string) => void
  navigation: ScanBarcodeScreenNavigationType
}
export const ScanBarcodeViewContainer: React.FC<Props> = ({
  isActive = true,
  onScan = () => {},
  navigation,
}) => {
  const { user } = useAuth()

  const handleScan = (barcode: string) => {
    if (!user) {
      throw new Error('This should never happen, as user is in the context')
    }
    saveBarcode(user.id, barcode)
    onScan(barcode)
  }

  return (
    <ScanBarcodeView
      onScan={handleScan}
      isActive={isActive}
      navigation={navigation}
    />
  )
}
