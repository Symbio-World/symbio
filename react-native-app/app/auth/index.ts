import { createAuthProvider } from './authContext'
import { signInAnonymously } from './signInAnonymously'
import { fetchEmail } from './fetchEmail'

export { useAuth, User } from './authContext'

export const AuthProvider = createAuthProvider({
  signInAnonymously,
  fetchEmail
})
