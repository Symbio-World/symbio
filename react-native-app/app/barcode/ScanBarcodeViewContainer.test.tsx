import 'react-native'
import * as React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import { ScanBarcodeViewContainer } from './ScanBarcodeViewContainer'
import { ScanBarcodeView } from './ScanBarcodeView'
import { useAuth } from '../auth'
import { saveBarcode } from './saveBarcode'

jest.mock('../auth')
jest.mock('./saveBarcode')

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => jest.fn(),
}))

describe('ScanBarcodeViewContainer', () => {
  const user = { id: 'id' }

  beforeEach(() => {
    ;(useAuth as jest.Mock).mockImplementation(() => ({ user }))
    ;(saveBarcode as jest.Mock).mockImplementation(() => Promise.resolve())
  })

  it('renders correctly', () => {
    const { toJSON } = render(<ScanBarcodeViewContainer />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('saves barcode', () => {
    const barcode = 'barcode'
    const { getByType } = render(<ScanBarcodeViewContainer />)
    fireEvent(getByType(ScanBarcodeView), 'onScan', barcode)
    expect(saveBarcode).toHaveBeenCalledWith(user.id, barcode)
  })

  it('triggers onScan callback', () => {
    const barcode = 'barcode'
    const handleScan = jest.fn()
    const { getByType } = render(
      <ScanBarcodeViewContainer onScan={handleScan} />,
    )
    fireEvent(getByType(ScanBarcodeView), 'onScan', barcode)
    expect(handleScan).toHaveBeenCalledWith(barcode)
  })
})
