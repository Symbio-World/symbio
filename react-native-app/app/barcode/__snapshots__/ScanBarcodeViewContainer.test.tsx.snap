// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanBarcodeViewContainer renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 64,
        "opacity": 1,
      }
    }
  >
    <Text>
      Trigger barcode
    </Text>
  </View>
  <BarcodePicker
    onBarcodeFrameAvailable={[Function]}
    onChangeTrackedCodes={[Function]}
    onPropertyChanged={[Function]}
    onRecognizeNewCodes={[Function]}
    onScan={[Function]}
    onSettingsApplied={[Function]}
    onTextRecognized={[Function]}
    onWarnings={[Function]}
    scanSettings={
      ScanSettings {
        "symbologies": Object {
          "code128": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "code39": SymbologySettings {
            "activeSymbolCounts": Array [
              7,
              8,
              9,
              10,
              11,
              12,
              13,
              14,
              15,
              16,
              17,
              18,
              19,
              20,
            ],
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "data-matrix": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "ean13": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "ean8": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "itf": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "qr": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "upca": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
          "upce": SymbologySettings {
            "checksums": Array [],
            "enabled": true,
            "extensions": Array [],
          },
        },
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "bottom": 0,
        },
        Object {
          "width": "100%",
        },
        Object {
          "height": 64,
        },
        Object {
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "#000",
        },
        Object {
          "opacity": 0.5,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
          },
          Object {
            "textAlign": "center",
          },
          Object {
            "fontWeight": "bold",
          },
          Object {
            "fontSize": 30,
          },
        ]
      }
    >
      Scan barcode
    </Text>
  </View>
</View>
`;
